protocols/autonatv2/src/request_response.rs

```diff
@@ -0,0 +1,274 @@
+                    .into_iter()
+                    .map(|e| e.to_vec())
+                    .map(|e| {
+                        Multiaddr::try_from(e).map_err(
+                            |err| new_io_invalid_data_err!(format!("invalid multiaddr: {}", err))
+                        )
+                    })
+                    .collect::<Result<Vec<_>, io::Error>>()?;
+                let nonce = check_existence!(nonce)?;
+                Ok(Self::Dial(DialRequest { addrs, nonce }))
+            }
+            proto::mod_Message::OneOfmsg::dialDataResponse(proto::DialDataResponse { data }) => {
+                let data_count = check_existence!(data)?.len();
+                Ok(Self::Data(DialDataResponse { data_count }))
+            }
+            _ => Err(new_io_invalid_data_err!(
+                "invalid message type, expected dialRequest or dialDataResponse"
+            )),
+        }
+    }
+
+    pub(crate) async fn write_into(self, io: impl AsyncWrite + Unpin) -> io::Result<()> {
+        let msg = match self {
+            Request::Dial(DialRequest { addrs, nonce }) => {
+                let addrs = addrs.iter().map(|e| e.to_vec()).collect();
+                let nonce = Some(nonce);
+                proto::Message {
+                    msg: proto::mod_Message::OneOfmsg::dialRequest(
+                        proto::DialRequest { addrs, nonce }
+                    ),
+                }
+            }
+            Request::Data(DialDataResponse { data_count }) => {
+                debug_assert!(
+                    data_count <= DATA_FIELD_LEN_UPPER_BOUND,
+                    "data_count too large"
+                );
+                static DATA: &[u8] = &[0u8; DATA_FIELD_LEN_UPPER_BOUND];
+                proto::Message {
+                    msg: proto::mod_Message::OneOfmsg::dialDataResponse(proto::DialDataResponse {
+                        data: Some(Cow::Borrowed(&DATA[..data_count])),
+                    }),
+                }
+            }
+        };
+        FramedWrite::new(io, Codec::<proto::Message>::new(REQUEST_MAX_SIZE))
+            .send(msg)
+            .await
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum Response {
+    Dial(DialResponse),
+    Data(DialDataRequest),
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct DialDataRequest {
+    pub(crate) addr_idx: usize,
+    pub(crate) num_bytes: usize,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct DialResponse {
+    pub(crate) status: proto::mod_DialResponse::ResponseStatus,
+    pub(crate) addr_idx: usize,
+    pub(crate) dial_status: proto::DialStatus,
+}
+
+impl Response {
+    pub(crate) async fn read_from(io: impl AsyncRead + Unpin) -> std::io::Result<Self> {
+        let msg = FramedRead::new(io, Codec::<proto::Message>::new(REQUEST_MAX_SIZE))
+            .next()
+            .await
+            .ok_or(io::Error::new(io::ErrorKind::UnexpectedEof, "eof"))??;
```

Comment:
```suggestion
            .ok_or(io::ErrorKind::UnexpectedEof)??;
```

Can be slightly simpler because `io::Error` implements `From<io::ErrorKind>`.

Reply:
That is changed.